airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StagingSqlOperations.java

```diff
@@ -0,0 +1,207 @@
+import io.airbyte.protocol.models.AirbyteRecordMessage;
+import java.io.File;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVPrinter;
+import org.apache.commons.lang3.NotImplementedException;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SnowflakeS3StagingSqlOperations extends SnowflakeSqlOperations implements StagingOperations {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SnowflakeSqlOperations.class);
+  private static final int DEFAULT_UPLOAD_THREADS = 10; // The S3 cli uses 10 threads by default.
+  private static final int DEFAULT_QUEUE_CAPACITY = DEFAULT_UPLOAD_THREADS;
+  private static final int DEFAULT_PART_SIZE = 10;
+  private static final int UPLOAD_RETRY_LIMIT = 3;
+
+  private final NamingConventionTransformer nameTransformer;
+  private final S3DestinationConfig s3Config;
+  private AmazonS3 s3Client;
+
+  public SnowflakeS3StagingSqlOperations(final NamingConventionTransformer nameTransformer,
+                                         final AmazonS3 s3Client,
+                                         final S3DestinationConfig s3Config) {
+    this.nameTransformer = nameTransformer;
+    this.s3Client = s3Client;
+    this.s3Config = s3Config;
+  }
+
+  @Override
+  public String getStageName(final String namespace, final String streamName) {
+    return nameTransformer.applyDefaultCase(String.join("_",
+        nameTransformer.convertStreamName(namespace),
+        nameTransformer.convertStreamName(streamName)));
+  }
+
+  @Override
+  public String getStagingPath(final String connectionId, final String namespace, final String streamName, final DateTime writeDatetime) {
+    // see https://docs.snowflake.com/en/user-guide/data-load-considerations-stage.html
+    return nameTransformer.applyDefaultCase(String.format("%s/%s/%s/%02d/%02d/%02d/",
+        connectionId,
+        getStageName(namespace, streamName),
+        writeDatetime.year().get(),
+        writeDatetime.monthOfYear().get(),
+        writeDatetime.dayOfMonth().get(),
+        writeDatetime.hourOfDay().get()));
+  }
+
+  @Override
+  public void uploadRecordsToStage(final JdbcDatabase database, final File dataFile, final String schemaName, final String path) throws Exception {
+    throw new NotImplementedException("placeholder function is not implemented yet");
+  }
+
+  @Override
+  public void insertRecordsInternal(final JdbcDatabase database,
+                                    final List<AirbyteRecordMessage> records,
+                                    final String schemaName,
+                                    final String path) {
+    final List<Exception> exceptionsThrown = new ArrayList<>();
+    boolean succeeded = false;
+    while (exceptionsThrown.size() < UPLOAD_RETRY_LIMIT && !succeeded) {
```

Comment:
An advantage of buffering outside of the `StreamTransferManager` is now that we can retry upload requests of small "buffered" parts if we run into network exceptions (which happened to me during my multiple benchmarks runs)...

Without this retry logic, the whole sync would be retried and if another small part fails, the full sync won't be able to complete